File: engine/tiles.py
"""Tile types and their static properties (walkable / vision / bonuses)."""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum


class TileType(str, Enum):
    PLAIN = "plain"
    LAKE = "lake"
    WALL = "wall"
    HIGH_GROUND = "high_ground"
    CAPTURE = "capture"


_CHAR_MAP = {
    ".": TileType.PLAIN,
    "~": TileType.LAKE,
    "#": TileType.WALL,
    "^": TileType.HIGH_GROUND,
    "*": TileType.CAPTURE,
}


def tile_from_char(ch: str) -> TileType:
    try:
        return _CHAR_MAP[ch]
    except KeyError as exc:
        raise ValueError(f"unknown map character: {ch!r}") from exc


@dataclass(frozen=True)
class TileProps:
    walkable: bool
    blocks_vision: bool
    is_high_ground: bool
    is_capture_point: bool


_PROPS: dict[TileType, TileProps] = {
    TileType.PLAIN: TileProps(True, False, False, False),
    TileType.LAKE: TileProps(False, False, False, False),
    TileType.WALL: TileProps(False, True, False, False),
    TileType.HIGH_GROUND: TileProps(True, False, True, False),
    TileType.CAPTURE: TileProps(True, False, False, True),
}


def props_of(tile: TileType) -> TileProps:
    return _PROPS[tile]

File: viewer/src/map/tiles.ts
/**
 * Tile types — mirrors engine/tiles.py.
 *
 * Map ASCII:
 *   .  → PLAIN
 *   ~  → LAKE      (impassable, does not block vision)
 *   #  → WALL      (impassable, blocks vision)
 *   ^  → HIGH_GROUND (passable, elevated)
 *   *  → CAPTURE   (passable, capture point)
 */

export const TileType = {
  PLAIN: 'plain',
  LAKE: 'lake',
  WALL: 'wall',
  HIGH_GROUND: 'high_ground',
  CAPTURE: 'capture',
} as const;

export type TileType = (typeof TileType)[keyof typeof TileType];

export interface TileProps {
  readonly walkable: boolean;
  readonly blocksVision: boolean;
  readonly isHighGround: boolean;
  readonly isCapturePoint: boolean;
}

const TILE_CHAR_MAP: Record<string, TileType> = {
  '.': TileType.PLAIN,
  '~': TileType.LAKE,
  '#': TileType.WALL,
  '^': TileType.HIGH_GROUND,
  '*': TileType.CAPTURE,
};

const TILE_PROPS: Record<TileType, TileProps> = {
  [TileType.PLAIN]:       { walkable: true,  blocksVision: false, isHighGround: false, isCapturePoint: false },
  [TileType.LAKE]:        { walkable: false, blocksVision: false, isHighGround: false, isCapturePoint: false },
  [TileType.WALL]:        { walkable: false, blocksVision: true,  isHighGround: false, isCapturePoint: false },
  [TileType.HIGH_GROUND]: { walkable: true,  blocksVision: false, isHighGround: true,  isCapturePoint: false },
  [TileType.CAPTURE]:     { walkable: true,  blocksVision: false, isHighGround: false, isCapturePoint: true  },
};

export function tileFromChar(ch: string): TileType {
  const t = TILE_CHAR_MAP[ch];
  if (t === undefined) {
    throw new Error(`unknown map character: ${JSON.stringify(ch)}`);
  }
  return t;
}

export function propsOf(tile: TileType): TileProps {
  return TILE_PROPS[tile];
}
